cppcache/integration-test-2/DataSerializableTest.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <geode/DataSerializable.hpp>
+#include <geode/RegionShortcut.hpp>
+#include <geode/DataInput.hpp>
+#include <geode/DataOutput.hpp>
+#include <geode/RegionFactory.hpp>
+#include <geode/TypeRegistry.hpp>
+
+#include "framework/Cluster.h"
+
+namespace {
+
+using namespace apache::geode::client;
+
+class Simple : public DataSerializable {
+ public:
+  inline Simple() : Simple("TestSimple", 30) {}
+
+  inline Simple(std::string name, int age)
+      : name_(std::move(name)), age_(age) {}
+
+  ~Simple() noexcept override = default;
+
+  using DataSerializable::fromData;
+  using DataSerializable::toData;
+
+  void fromData(DataInput& dataInput) override {
+    name_ = dataInput.readString();
+    age_ = dataInput.readInt32();
+  }
+
+  void toData(DataOutput& dataOutput) const override {
+    dataOutput.writeString(name_);
+    dataOutput.writeInt(age_);
+  }
+
+  static std::shared_ptr<DataSerializable> createDeserializable() {
+    return std::make_shared<Simple>();
+  }
+
+  std::string getName() { return name_; }
+  int getAge() { return age_; }
+
+ private:
+  std::string name_;
+  int age_;
+};
+
+class DataSerializableObject : public DataSerializable {
+ public:
+  inline DataSerializableObject()
+      : DataSerializableObject("TestDataSerializableObject", nullptr, nullptr) {
+  }
+
+  inline DataSerializableObject(std::string name,
+                                std::shared_ptr<CacheableStringArray> csArray,
+                                std::shared_ptr<Simple> simple)
+      : name_(std::move(name)), csArray_(csArray), simple_(simple) {}
+
+  ~DataSerializableObject() noexcept override = default;
+
+  using DataSerializable::fromData;
+  using DataSerializable::toData;
+
+  void fromData(DataInput& dataInput) override {
+    name_ = dataInput.readString();
+    csArray_ =
```

Comment:
Why the ugliness of using an array of strings?

Reply:
We originally thought we had a serialization problem related to CacheableStringArray, which is one of the types in our Portfolio object. The only tests using CacheableStringArray were Cq tests (which are marked as FLAKY), so we wrote this very simple test.

Reply:
There are a ton of serialization tests. You are telling me that none of the PDX or DataSerializable tests have a string array in them?

Reply:
you might be surprised...

Reply:
*facepalm*